docs/sphinx/source/contributing.rst

```diff
@@ -81,29 +81,113 @@ changes, such as fixing documentation typos.
+and ModelChain. Contributors will need to add tests that correspond to
+the layer that they modify.
+
+Functions
+---------
+Tests of core pvlib functions should ensure that the function returns
+the desired output for a variety of function inputs. The tests should be
+independent of other pvlib functions (see :issue:`394`). The tests
+should ensure that all reasonable combinations of input types (floats,
+nans, arrays, series, scalars, etc) work as expected. Remember that your
+use case is likely not the only way that this function will be used, and
+your input data may not be generic enough to fully test the function.
+Write tests that cover the full range of validity of the algorithm.
+It is also important to write tests that assert the return value of the
+function or that the function throws an exception when input data is
+beyond the range of algorithm validity.
+
+PVSystem/Location
+-----------------
+The PVSystem and Location classes provide convenience wrappers around
+the core pvlib functions. The tests in test_pvsystem.py and
+test_location.py should ensure that the method calls correctly wrap the
+function calls. Many PVSystem/Location methods pass one or more of their
+object's attributes (e.g. PVSystem.module_parameters, Location.latitude)
+to a function. Tests should ensure that attributes are passed correctly.
+These tests should also ensure that the method returns some reasonable
+data, though the precise values of the data should be covered by
+function-specific tests discussed above.
+
+We prefer to use the ``pytest-mock`` framework to write these tests. The
+test below shows an example of testing the ``PVSystem.ashraeiam``
+method. ``mocker`` is a ``pytest-mock`` object. ``mocker.spy`` adds
+features to the ``pvsystem.ashraeiam`` *function* that keep track of how
+it was called. Then a ``PVSystem`` object is created and the
+``PVSystem.ashraeiam`` *method* is called in the usual way. The
+``PVSystem.ashraeiam`` method is supposed to call the
+``pvsystem.ashraeiam`` function with the angles supplied to the method
+call and the value of ``b`` that we defined in ``module_parameters``.
+The ``pvsystem.ashraeiam.assert_called_once_with`` call will test this
+for us! Finally, we check that the output of the method call is
+reasonable.
+
+.. code-block:: python
+    def test_PVSystem_ashraeiam(mocker):
+        # mocker is a pytest-mock object.
+        # mocker.spy adds code to a function to keep track of how its called
+        mocker.spy(pvsystem, 'ashraeiam')
+
+        # set up inputs
+        module_parameters = pd.Series({'b': 0.05})
+        system = pvsystem.PVSystem(module_parameters=module_parameters)
+        thetas = 1
+
+        # call the method
+        iam = system.ashraeiam(thetas)
+
+        # did the method call the function as we expected?
+        # mocker.spy added assert_called_once_with to the function
+        pvsystem.ashraeiam.assert_called_once_with(thetas, b=0.05)
```

Comment:
In the example above, either ``b=0.05`` or ``b=module_parameters['b']`` would work. What is important is that something dict-like is passed to ``PVSystem``. For example, this would also work:  
``system = pvsystem.PVSystem(module_parameters={'b': 0.05})`` followed by the assert statement as shown. Or this would work:

```
b = 0.05
system = pvsystem.PVSystem(module_parameters=dict(b=b))
# ...
pvsystem.ashraeiam.assert_called_once_with(thetas, b=b)
```

Would one of the above be a better example? Or more comments?

Reply:
I think it would appear more clear to me if we change line 157 to `module_parameters = {'b' : 0.05}` which avoids recasting the dict as a pandas object.  Also, to me it is more consistent to use the dict in the assertion,  `assert_called_once_with(thetas, module_parameters['b'])` instead of `assert_called_once_with(thetas, b=0.05)`